### ISSUE TITLE
[BUG]: Matmul fails with large shapes

### ISSUE THREAD

**Issue by bohnstingl** [torch-spyre/torch-spyre#1918]
### What happened?

In vLLM, the `ParallelLMHead` uses a matmul with a relatively large size (x: `torch.Size([1, 4096])` and w: `torch.Size([49216, 4096])`) and this leads to issues with torch-spyre. 

### What did you expect to happen?

Matmul should be agnostic to the specific shapes of the inputs and outputs, just like on `cpu`.

### How can we reproduce it?

Run the script below

```
"""Minimal repro: dxp_standalone SIGABRT on lm_head-shaped F.linear.

The Spyre hardware compiler crashes when compiling a standalone
F.linear with vocab-sized output dimension (49216).

Shapes (from Granite 3.2 2b):
    x:      [1, 4096]      (single token, hidden_dim)
    weight: [49216, 4096]   (vocab_size, hidden_dim)
    output: [1, 49216]      (single token, vocab_size)

The same matmul compiles fine for typical hidden-dim-scale shapes
(e.g. [1, 4096] x [4096, 4096]^T) used by in-model linear layers.

Usage:
    python lm_head_matmul_repro.py
"""

import torch
import torch.nn.functional as F

HIDDEN_DIM = 4096
VOCAB_SIZE = 49216

def lm_head_matmul(x, weight):
    return F.linear(x, weight)

# --- Allocate on CPU, transfer to Spyre ---
x_cpu = torch.randn(1, HIDDEN_DIM, dtype=torch.float16)
w_cpu = torch.randn(VOCAB_SIZE, HIDDEN_DIM, dtype=torch.float16)

x_spyre = x_cpu.to("spyre")
w_spyre = w_cpu.to("spyre")

# --- This works: hidden-dim-scale matmul ---
print(f"Testing hidden-dim-scale matmul: x=[1, {HIDDEN_DIM}] @ w=[{HIDDEN_DIM}, {HIDDEN_DIM}]^T")
w_small_cpu = torch.randn(HIDDEN_DIM, HIDDEN_DIM, dtype=torch.float16)
w_small_spyre = w_small_cpu.to("spyre")
out_small = torch.compile(lm_head_matmul, dynamic=False)(x_spyre, w_small_spyre)
print(f"  OK — output shape: {out_small.shape}, device: {out_small.device}")

# --- This crashes: vocab-scale matmul ---
print(f"\nTesting vocab-scale matmul: x=[1, {HIDDEN_DIM}] @ w=[{VOCAB_SIZE}, {HIDDEN_DIM}]^T")
print("  Expecting dxp_standalone SIGABRT...")
out = torch.compile(lm_head_matmul, dynamic=False)(x_spyre, w_spyre)
print(f"  OK — output shape: {out.shape}, device: {out.device}")
```

### Any environmental details we need to know?

I am on these branches

```
DIR                  | BRANCH          | WT_STATE                     | LOCAL_SHORT  | LOCAL_FULL                               | EXPECT_BRANCH   | EXPECT_HASH                             |
flex                 | main            | CLEAN                        | d19074e5     | d19074e5c02f65c45d22aed344441896e13c7915 | main            | 5ad7fca99a2bc13513d42fa4957d7cd67a225067|
flex/common          | DETACHED_HEAD   | CLEAN                        | e708de6      | e708de6edbb75e98f2d81c1925640e6a60130069 | main            | 8a141d5f47c7e77204185eacc0fef4deb6fedb77|
deeptools            | master          | CLEAN                        | 7fe40271bf   | 7fe40271bf9fe043f72ab88f560f69b6958995fa | master          | 7fe40271bf9fe043f72ab88f560f69b6958995fa|
torch-spyre          | main            | CLEAN                        | 9d31e2d      | 9d31e2dab32cc36eeb4711d07613c131e9111ba8 | main            | 3652f3dd20a37de63eef8b3c3c5e6662e19fef56|
torch-spyre-docs     | main            | UNSTAGED                     | 5513731      | 551373118e9e16bab3c995723beb3aa70117c929 | main            | 551373118e9e16bab3c995723beb3aa70117c929|
```

### Anything else we need to know?

Is this issue related to https://github.com/torch-spyre/torch-spyre/issues/952 or https://github.com/torch-spyre/torch-spyre/issues/710?

cc @dgrove-oss @kiszk

### Relevant log output

```Shell

```

**Comment by bohnstingl** [comment]
Update, the repro script above fails with `VOCAB_SIZE > 4096`

**Comment by nayana-kumari** [comment]
The ParallelLMHead wrapping is working , forward_oot() is triggered correctly and the matmul reaches the Spyre compilation path.

the execution fails with a hardware/compiler error:

[CRITICAL] Cannot satisfy minimum split requirement for d0:
need 769 splits but only 16 cores remaining  
DtException: EAR overflow detected  
dxp_standalone died with SIGABRT

**Comment by bohnstingl** [comment]
This issue is related to https://github.com/torch-spyre/torch-spyre/issues/1392 and https://github.com/torch-spyre/torch-spyre/issues/1379

**Comment by umacdevi** [comment]
torch.matmul fails for the following pairs of operands. 

A = torch.randn(8, 64, 576, dtype=torch.float16)
B = torch.randn(8, 576, 15876, dtype=torch.float16

A = torch.randn(1, 1024, 588, dtype=torch.float16)
B = torch.randn(1, 588, 196701, dtype=torch.float16)

A = torch.randn(8, 64, 75, dtype=torch.float16)
B = torch.randn(8, 75, 16384, dtype=torch.float16)

A = torch.randn(8, 8, 27, dtype=torch.float16)
B = torch.randn(8, 27, 4096, dtype=torch.float16

This was found while implementing conv2d via unfold and matmul. See https://github.com/torch-spyre/torch-spyre/issues/444#issuecomment-4534774636.

**Comment by umacdevi** [comment]
The above matmuls are working in my build now. They were due to some local changes that led to sdsc differences.

**Comment by bohnstingl** [comment]
For me the matmul with the bigger shapes still fails

```
Testing hidden-dim-scale matmul: x=[1, 4096] @ w=[4096, 4096]^T
/home/boh/dt-inductor2/pytorch/torch/fx/interpreter.py:180: TqdmMonitorWarning: tqdm:disabling monitor support (monitor_interval = 0) due to:
can't start new thread
  pbar = tqdm(
  OK — output shape: torch.Size([1, 4096]), device: spyre:0

Testing vocab-scale matmul: x=[1, 4096] @ w=[49216, 4096]^T
  Expecting dxp_standalone SIGABRT...
[WARNING] [work_division] Cannot satisfy span limit for tensor buf1: coord=floor(d0/64) still evaluates to 769 after splits. Inner dimensions cannot reduce span further. Best span=403177472, limit=268435456.
[WARNING] [work_division] Cannot satisfy span limit for tensor buf1: coord=floor(d0/64) still evaluates to 769 after splits. Inner dimensions cannot reduce span further. Best span=403177472, limit=268435456.
[CRITICAL] [work_division] buf1: per-core tensor span 384.50 MB (shape=[49216, 4096], dtype=torch.float16, device_size=[769, 4096, 64], splits={d0: 1, d1: 2}) exceeds hardware limit of 256.00 MB
[CRITICAL] [work_division] buf0: per-core tensor span 384.50 MB (shape=[49216, 4096], dtype=torch.float16, device_size=[769, 4096, 64], splits={d0: 1, d1: 32}) exceeds hardware limit of 256.00 MB
terminate called after throwing an instance of 'DtException'
  what():  DtException: EAR overflow detected, file /home/boh/dt-inductor2/deeptools/dcc/src/Transform/Dataflow/MutableAddrSplitting.cpp line 780
```

I am on versions:
```
DIR                  | BRANCH          | WT_STATE                     | LOCAL_SHORT  | LOCAL_FULL                               | EXPECT_BRANCH   | EXPECT_HASH                             |
flex                 | main            | CLEAN                        | 8f06481d     | 8f06481de71d9cf28480f88bb7e8e02d60a75bda | main            | 908bdcc07b51e56aa8f45a0bcececfb2e59ababe|
flex/common          | main            | CLEAN                        | 8a141d5      | 8a141d5f47c7e77204185eacc0fef4deb6fedb77 | main            | 8a141d5f47c7e77204185eacc0fef4deb6fedb77|
deeptools            | master          | CLEAN                        | b7f6fdbfa7   | b7f6fdbfa7275a493fe62ac128b058a9d99d45c3 | master          | 940d8d34fe31c1994d0e446ba1a7272795ac8837|
```

cc @GOavi101, @R3hankhan123  I think this is related to the issue https://github.com/torch-spyre/spyre-inference/issues/212, right?

**Comment by R3hankhan123** [comment]
Yes @bohnstingl

**Comment by bohnstingl** [comment]
@dgrove-oss @tardieu I see that several similar issues are being closed recently, but I tested the repro script above with commit `89158fc144c8b69a38bc57c91d9e43a988f158e9` and it still fails with
```
Testing vocab-scale matmul: x=[1, 4096] @ w=[49216, 4096]^T
  Expecting dxp_standalone SIGABRT...
[WARNING] [work_division] Cannot satisfy span limit for tensor buf1: coord=floor(d0/64) still evaluates to 769 after splits. Inner dimensions cannot reduce span further. Best span=403177472, limit=268435456.
[WARNING] [work_division] Cannot satisfy span limit for tensor buf1: coord=floor(d0/64) still evaluates to 769 after splits. Inner dimensions cannot reduce span further. Best span=403177472, limit=268435456.
[CRITICAL] [work_division] buf1: per-core tensor span 384.50 MB (shape=[49216, 4096], dtype=torch.float16, device_size=[769, 4096, 64], splits={d0: 1, d1: 2}) exceeds hardware limit of 256.00 MB
[CRITICAL] [work_division] buf0: per-core tensor span 384.50 MB (shape=[49216, 4096], dtype=torch.float16, device_size=[769, 4096, 64], splits={d0: 1, d1: 32}) exceeds hardware limit of 256.00 MB
terminate called after throwing an instance of 'DtException'
  what():  DtException: EAR overflow detected
```

**Comment by dgrove-oss** [comment]
Thanks for verifying the bug still exists.

This will be addressed by the work @Hemaprasannakc is leading to enable working set reduction based on tensor span even if the user hasn't annotated the program with spyre_hints to specify the reduction to perform.

**Comment by bohnstingl** [comment]
@dgrove-oss thank you for clarifying.

@Hemaprasannakc is there a way that we can use `spyre_hints` to resolve the issue from above? How would you use them in this case?

**Comment by Hemaprasannakc** [comment]
@bohnstingl  , Initial approach will not be covering `spyre_hints` immediately , However eventually we are planning to integrate them as we progress . Please refer to https://github.com/torch-spyre/torch-spyre/issues/206

**Comment by mudhakar** [comment]
This looks like standard granite LM head size. And it works today without turning on working set right sizing. The intra-op tiling in matmul is sufficient to handle this. 

working set right sizing is required for inter-op tiling (such as MLP and attention). @dgrove-oss , @ani300 ?

**Comment by bohnstingl** [comment]
@mudhakar unfortunately, I cannot comment on the working set right sizing, but the repro script above still fails on the commit `7cb77e40ac98f4c93d2a9f584b748538f2ef8df8`.

**Comment by Hemaprasannakc** [comment]
#3197  #3198 -> Feature is related to solve the end to end auto tiling .

**Comment by cyang49** [comment]
@Hemaprasannakc could you share a high-level description of the proposed solution to this? Thank you

**Comment by mudhakar** [comment]
**Root-caused and confirmed empirically on latest `main` (`torch-spyre e4cd21e7`, full stack rebuilt).**

This is a **single-op span-reduction** problem, not a working-set / inter-op problem. No WSR is required to fix it.

**Root cause.** The weight `[49216, 4096]` gets device layout `[769, 4096, 64]`. In `work_division`, per-core span is `outer_range × Π(full inner device_size) × itemsize`, so the span is set by the *outermost* varying dim times the *full* inner stride. Splitting an inner dim (K=4096) therefore does **not** reduce the span — only splitting the outer stick dim `d0` does, and that split must divide the stick count.

`49216 / 64 = 769`, and **769 is prime** → the only splits of `d0` are `{1, 769}`, neither of which fits in ≤32 cores. Span stays at 384.5 MB > 256 MB → EAR overflow → `dxp_standalone` SIGABRT.

**Empirical A/B (latest `main`):**

| vocab | sticks (÷64) | factorization | result |
|---|---|---|---|
| 49216 | 769 | **prime** | **FAIL** — `Cannot satisfy span limit … floor(d0/64) still evaluates to 769`; CRITICAL `buf1 splits={d0:1,d1:2}`, `buf0 {d0:1,d1:32}`, 384.5 MB > 256 MB; `dxp_standalone died with SIGABRT` |
| 49280 | 770 = 2·5·7·11 | composite | **PASS** — `out.shape=(1, 49280)`, **no span warnings** (span reduction splits `d0` cleanly, e.g. ÷2 → 385 sticks → 192.5 MB) |

**Fix.** Pad the weight's outer stick-count to the nearest composite value — thanks to @cyang49 for the idea that 769→770 is enough (**+1 stick, ~0.13% memory**). The existing span-reduction pass then handles it with no new machinery. The padded logits are garbage and must be narrowed/masked before argmax/sampling. Uneven/remainder coarse tiling (`[512, 127]`) would avoid the overhead entirely but is not supported today.

cc @dgrove-oss @Hemaprasannakc @cyang49

**Comment by Hemaprasannakc** [comment]
@mudhakar , @cyang49 ,

For this specific repro (`single token, x=[1,4096]`)  Weight's span doesn't depend on token count at all, so if span_reduction can handle prime divisibility  (`with the 769->770 pad)` fully covers this case .

Where this connects to a related, currently-separate problem of [3203](https://github.com/torch-spyre/torch-spyre/issues/3203)-  output does scale with token count, and if a large enough batch/prefill ever pushes the *output* tensor itself past the span limit, tiling its vocab dimension would go through coarse-tiling (manual spyre_hint or our automatic span-overflow pass), not span_reduction. 

Vocab is still 64*769 either way, so that tiling attempt would hit the exact same prime wall -- and we've already confirmed our own pass fails on it the same way (`Cannot auto-tile bufX: no combined split..`.) that needs to be handled in similar way of padding fix.

**Comment by Hemaprasannakc** [comment]
@mudhakar , @cyang49 

Also Quick follow-up on where this lands ....769->770 pad implemented as an early, graph-level resize of weight (similar to how insert_bmm_padding pads K today, before coarse-tiling ever sees the op) -- or is it internal to span_reduction's own logic specifically?

If it's the former,  automatic span-overflow pass would likely see the already-padded size and just work, no changes needed on it's end.
 
If it's the latter, it'd need it's own version for the coarse-tiling path (reusing the same padding pattern, but also needing a slice-back step to strip the padded/garbage columns from the output before use, since weight and output share the vocab dimension here). 

Please confirm.

**Comment by cyang49** [comment]
@Hemaprasannakc the padding solution is theoretical, I don't think it's being implemented anywhere. Thus I was asking you what was the solution you proposed.

@dgrove-oss please comment on whether we want to use padding as the solution?

**Comment by cyang49** [comment]
@Hemaprasannakc also if WSR allows a [768, 1] or [512, 127] split on the 769 sized dimension it can also solve it.

**Comment by Hemaprasannakc** [comment]
> [@Hemaprasannakc](https://github.com/Hemaprasannakc) also if WSR allows a [768, 1] or [512, 127] split on the 769 sized dimension it can also solve it.

@cyang49 
Two paths depending on how the theoretical padding discussion  above lands:

**If padding works out (769→770, composite)**: span_reduction can already handle it, and WSR doesn't need to change at all — the padded dimension gets a normal exact-divisor split.

**If padding isn't pursued** — pessimistically assuming span_reduction can't be made to handle this — then WSR would probably need to pick up the slack itself, similar to how we already generate hints for large spans that are evenly divisible today. 

Right now that flow is: find an exact stick-aligned divisor for the dimension, emit a DimHint with that split_count, and let coarse-tile stamp uniform per-tile loop info from it. For a prime dimension like 769 there's no such divisor, so nothing gets emitted and the op is left unsupported


**One direction worth exploring:** extend the DimHint/coarse-tile representation from a single split_count to an explicit list of tile sizes — e.g. N stick-aligned tiles plus one smaller remainder tile, along the lines of [768, 1] / [512, 257] suggestion — instead of requiring every tile to be the same size. The remainder tile would need special-casing in the loop/address codegen path, since today's per-tile stride and offset computation assumes every tile in the loop is identical.

**Comment by dgrove-oss** [comment]
Padding to optimize for core division is one of the many things in the inductor backlog (https://github.com/torch-spyre/torch-spyre/issues/741).   It's relatively low priority because there are source-level workarounds available to the programmer (for example using a `to` with layout where the STL specifies additional padding).

**Comment by Hemaprasannakc** [comment]
Thanks @dgrove-oss, 

To make sure I understand the workaround correctly: for something like our vocab=49216 case, could the caller pad the tensor's layout up front (e.g. to 770 sticks) via the STL mechanism you mentioned, before it ever reaches WSR/span_reduction? If so, that sounds like it already solves the immediate problem without needing either compiler-side padding or the remainder-tile . we could just document that as the recommended workaround for primes like this, and treat it as genuinely unsupported (with that pointer) rather than building new infra for a low-priority case.

Could you point me to an example of using that layout/STL padding, if one exists? Want to make sure I'm recommending the right pattern before I change current code from "unsupported" .

**Comment by mudhakar** [comment]
@Hemaprasannakc , @dgrove-oss : 
Upfront padding for this use case is supported din HF Adapters: ParallelLMHead adapter.  It pads the vocab dim and masks the padded logits.

**Comment by mudhakar** [comment]
https://github.com/torch-spyre/hf-adapters/blob/main/hf_adapters/hf_common.py#L752

**Comment by mudhakar** [comment]
Root cause + fix location (vLLM-reported case):

The lm_head weight sticks on the vocab dim; when `vocab/64` has a large prime factor (`49216/64 = 769`), work division can't split the stick axis and the per-core span exceeds the 256 MB EAR limit → SIGABRT. Analysis in #3217.

Not fixable in torch-spyre today (no uneven-tile support; on-device padding re-incurs the span). The fix is **host-side**: pad the lm_head vocab to a *smooth* stick count before it goes to device. The HF-adapters path does this via `pad_lm_head()` (`hf_adapters/hf_common.py`). The **vLLM pathway needs the equivalent** — round `vocab/64` up until `largest_prime_factor × hidden × 64 × dtype_bytes ≤ 256 MB` (bumps are 0–1 sticks). Requesting that on the vLLM / spyre-inference side.

cc @bohnstingl @dgrove-oss
